### ISSUE TITLE
Supply class-level dartdocs when no constructor-level dartdoc is defined for getHovers in analyzer

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#28537]
Raised by @eseidel in https://github.com/Dart-Code/Dart-Code/issues/240. I think this would be best done in the analyzer if you think it's worthwhile (and would also then work for all other editors). I think the reason the first one works is that it just looks like a reference to the class rather than the constructor. I agree that the second one would be best falling back to the class dart doc because the constructor doesn't have one.

-----

**@eseidel**:

Hovering a new using a constructor shows the class docs:

![constructor_hover](https://cloud.githubusercontent.com/assets/478541/21473342/c7d562d2-cabb-11e6-9fef-bbe4508e6478.png)

But not hovering an annotation (which uses the same constructor):

![annotation_hover](https://cloud.githubusercontent.com/assets/478541/21473335/85a24740-cabb-11e6-8a11-15b794ced5f6.png)

No clue if the missing functionality here is in the plugin or the analyzer or what. :)

**Comment by marcoms** [comment]
I think a case can be made that class docs could be included even if constructor docs _are_ present.

Frequently in Flutter there are very helpful and comprehensive class docs, but the constructor just provides a couple sentences of description. Example:

```dart
/// A convenience widget that combines common painting, positioning, and sizing
/// widgets.
///
/// A container first surrounds the child with [padding] (inflated by any
/// borders present in the [decoration]) and then applies additional
/// [constraints] to the padded extent (incorporating the `width` and `height`
/// as constraints, if either is non-null). The container is then surrounded by
/// additional empty space described from the [margin].
///
/// During painting, the container first applies the given [transform], then
/// paints the [decoration] to fill the padded extent, then it paints the child,
/// and finally paints the [foregroundDecoration], also filling the padded
/// extent.
///
/// Containers with no children try to be as big as possible unless the incoming
/// constraints are unbounded, in which case they try to be as small as
/// possible. Containers with children size themselves to their children. The
/// `width`, `height`, and [constraints] arguments to the constructor override
/// this.
///
/// ## Layout behavior
///
/// _See [BoxConstraints] for an introduction to box layout models._
///
/// Since [Container] combines a number of other widgets each with their own
/// layout behavior, [Container]'s layout behavior is somewhat complicated.
///
/// tl;dr: [Container] tries, in order: to honor [alignment], to size itself to
/// the [child], to honor the `width`, `height`, and [constraints], to expand to
/// fit the parent, to be as small as possible.
///
/// More specifically:
///
/// If the widget has no child, no `height`, no `width`, no [constraints],
/// and the parent provides unbounded constraints, then [Container] tries to
/// size as small as possible.
///
/// If the widget has no child and no [alignment], but a `height`, `width`, or
/// [constraints] are provided, then the [Container] tries to be as small as
/// possible given the combination of those constraints and the parent's
/// constraints.
///
/// If the widget has no child, no `height`, no `width`, no [constraints], and
/// no [alignment], but the parent provides bounded constraints, then
/// [Container] expands to fit the constraints provided by the parent.
///
/// If the widget has an [alignment], and the parent provides unbounded
/// constraints, then the [Container] tries to size itself around the child.
///
/// If the widget has an [alignment], and the parent provides bounded
/// constraints, then the [Container] tries to expand to fit the parent, and
/// then positions the child within itself as per the [alignment].
///
/// Otherwise, the widget has a [child] but no `height`, no `width`, no
/// [constraints], and no [alignment], and the [Container] passes the
/// constraints from the parent to the child and sizes itself to match the
/// child.
///
/// The [margin] and [padding] properties also affect the layout, as described
/// in the documentation for those properties. (Their effects merely augment the
/// rules described above.) The [decoration] can implicitly increase the
/// [padding] (e.g. borders in a [BoxDecoration] contribute to the [padding]);
/// see [Decoration.padding].
///
/// {@tool sample}
///
/// This example shows a 48x48 green square (placed inside a [Center] widget in
/// case the parent widget has its own opinions regarding the size that the
/// [Container] should take), with a margin so that it stays away from
/// neighboring widgets:
///
/// ```dart
/// Center(
///   child: Container(
///     margin: const EdgeInsets.all(10.0),
///     color: const Color(0xFF00FF00),
///     width: 48.0,
///     height: 48.0,
///   ),
/// )
/// ```
/// {@end-tool}
/// {@tool sample}
///
/// This example shows how to use many of the features of [Container] at once.
/// The [constraints] are set to fit the font size plus ample headroom
/// vertically, while expanding horizontally to fit the parent. The [padding] is
/// used to make sure there is space between the contents and the text. The
/// `color` makes the box teal. The [alignment] causes the [child] to be
/// centered in the box. The [foregroundDecoration] overlays a nine-patch image
/// onto the text. Finally, the [transform] applies a slight rotation to the
/// entire contraption to complete the effect.
///
/// ```dart
/// Container(
///   constraints: BoxConstraints.expand(
///     height: Theme.of(context).textTheme.display1.fontSize * 1.1 + 200.0,
///   ),
///   padding: const EdgeInsets.all(8.0),
///   color: Colors.teal.shade700,
///   alignment: Alignment.center,
///   child: Text('Hello World', style: Theme.of(context).textTheme.display1.copyWith(color: Colors.white)),
///   foregroundDecoration: BoxDecoration(
///     image: DecorationImage(
///       image: NetworkImage('https://www.example.com/images/frame.png'),
///       centerSlice: Rect.fromLTRB(270.0, 180.0, 1360.0, 730.0),
///     ),
///   ),
///   transform: Matrix4.rotationZ(0.1),
/// )
/// ```
/// {@end-tool}
///
/// See also:
///
///  * [AnimatedContainer], a variant that smoothly animates the properties when
///    they change.
///  * [Border], which has a sample which uses [Container] heavily.
///  * [Ink], which paints a [Decoration] on a [Material], allowing
///    [InkResponse] and [InkWell] splashes to paint over them.
///  * The [catalog of layout widgets](https://flutter.io/widgets/layout/).
class Container extends StatelessWidget {
  /// Creates a widget that combines common painting, positioning, and sizing widgets.
  ///
  /// The `height` and `width` values include the padding.
  ///
  /// The `color` argument is a shorthand for `decoration: new
  /// BoxDecoration(color: color)`, which means you cannot supply both a `color`
  /// and a `decoration` argument. If you want to have both a `color` and a
  /// `decoration`, you can pass the color as the `color` argument to the
  /// `BoxDecoration`.
```

Since users mostly reference the constructor, in order to access the full documentation, you basically have to Alt+Click and view the source code of the widget. IMO not the best DX.

**Comment by DanTup** [comment]
Showing class dartdocs always also came up at https://github.com/flutter/flutter/issues/96332.

I wonder whether the issue is more general than constructors though - if you changed a constructor to a static method, would seeing the dartdoc have less value, or is the desire really to see the docs of the return type of all methods/functions?

I also wonder if the same should go for code completion. Although it seems like in VS Code that's already showing the class dartdocs (instead of the constructor) here, and I'm not sure why (I can't reproduce in Android Studio). Seems like they should probably be consistent - although fixing that without also doing this might seem like a step backwards if there is value in seeing the class docs there.

**Comment by bwilkerson** [comment]
It seems like there are at least three issues here that would be worth thinking about separately.

1. Does it makes sense to display the class' documentation when a constructor without documentation is being referenced?

2. Should the class' documentation always be shown when hovering over a reference to a constructor.

3. What documentation should be displayed for code completion.

It seems to me that one guiding principle for answering these questions (and any question related to how documentation is displayed in the IDE) is to be consistent with dartdoc wherever it makes sense to do so.

So for the first issue, I'd say that it doesn't make sense to display the class' documentation for an undocumented constructor because dartdoc doesn't do so. I think it also makes sense to not do so just from the perspective that the documentation for the class isn't describing the constructor.

But the second issues is a bit different. I wonder whether it's really an observation that without type annotations there's no easy affordance to access the class' documentation. For example, if you wrote
```dart
Container c = Container();
```
then you could hover over the first occurrence of the class name to get the class' documentation and the second occurrence to get the constructor's documentation. Whereas if the first occurrence is replaced by `var`, then there is no hover help. It might we'll be worthwhile considering possible affordances for accessing the class' documentation in such situations.

For the third issue, I think we should always display the documentation for the element that would be referenced if the suggestion is accepted. If that's not currently the case, then I think we have a bug and we should fix it.

**Comment by DanTup** [comment]
>  It might we'll be worthwhile considering possible affordances for accessing the class' documentation in such situations.

I can't think of many ways we could do this besides adding the class documentation to the bottom of the constructor docs. I think the same may apply to return values of non-constructor functions too:

```dart
Foo foo = getFoo(); // Can easily access Foo class docs
final foo = getFoo(); // Cannot
```

Although accepting that also makes me think `final a = class.theFoo` isn't much different and you might want to see docs about the type of `theFoo`. This might make the solution more general (always show the class docs for the return type) - although given an issue in VS Code where it shows diagnostics right at the bottom of the docs (https://github.com/microsoft/vscode/issues/73120) I think adding more docs into tooltips there could be bad for usability until that's fixed (it's been open almost 3 years so I'm not hopeful for a fix in the near term).


> For the third issue, I think we should always display the documentation for the element that would be referenced if the suggestion is accepted. If that's not currently the case, then I think we have a bug and we should fix it.

I had a quick look at this, and believe it's because of how we fetch the documentation lazily for LSP with suggestion sets. We send two completions to the client (`MyClass` and `MyClass()` for the class/constructor respectively), but when we "resolve" them to get the `import` edit and documentation, we only have `MyClass` to look the element up (via `requestedLibraryElement.exportNamespace.get(requestedName)` so both the class and the constructor get the same thing.

I think this problem might go away when migrating to the `NotImportedContributor`. I filed an issue at https://github.com/Dart-Code/Dart-Code/issues/3787 as a reminder for me to check this once that work is done.

**Comment by bwilkerson** [comment]
> > It might we'll be worthwhile considering possible affordances for accessing the class' documentation in such situations.
>
> I can't think of many ways we could do this besides adding the class documentation to the bottom of the constructor docs. I think the same may apply to return values of non-constructor functions too:

A couple of brainstorming ideas:

- Could we add "links" of some kind to the hover text so that users could navigate to the docs for the type?

- We could definitely add a hover for `var`, `final`, or `const` that would display the docs for the inferred type (or explicit type if one is given).

But I definitely agree that appending the class docs to the end of the other might be a poor UX, especially given the size of the docs for most of the Flutter classes. Hence the desire to find a better way to make this information accessible to users. (And there doesn't have to be a single way if multiple ways make sense.)

**Comment by DanTup** [comment]
> Could we add "links" of some kind to the hover text so that users could navigate to the docs for the type?

For VS Code, we could append hyperlinks to the end of the markdown that go to the docs website - though I don't think we can do anything fancy like update the hover content to show another items docs. To do the first, we'd need to be able to construct URLs to get to those docs. This is a request that's been asked for before though, and I do often find myself opening my browser and going to pub.dev/packages/foo and then clicking the API documentation link on the side.


> We could definitely add a hover for var, final, or const that would display the docs for the inferred type (or explicit type if one is given).

That may help some, although not in places like the Flutter one where this came up recently:

```dart
final a = Container(
  child: ElevatedButton() // Want to see ElevatedButton docs by hovering on this, or when code-completing it
);
```

Right now you only see:

>  /// Create an ElevatedButton.
>  ///
>  /// The [autofocus] and [clipBehavior] arguments must not be null.

<img width="733" alt="Screenshot 2022-01-19 at 15 39 51" src="https://user-images.githubusercontent.com/1078012/150164718-fde19eed-402b-430e-a0d5-787d76660989.png">

**Comment by bwilkerson** [comment]
That's definitely a harder case. I'm not seeing anywhere obvious where a user could reasonably expect to be able to hover to get those docs. But if we could link to the class' docs externally (or even the constructor's docs because it ought to be easy to get from there to the docs for the class) that might help.

**Comment by DanTup** [comment]
I think having a hyperlink appended to the dartdoc that has "View docs on pub.dev" or similar would be good. I don't know how simple it is to build the correct URL though (I think it would need to be conditional on it being a package from pub and not just local, and take the version number into account?).

**Comment by bwilkerson** [comment]
Yes, it probably would. Flutter documentation might need to be handled differently, and in fact might want to be the first case we handle.

**Comment by FMorschel** [comment]
> So for the first issue, I'd say that it doesn't make sense to display the class' documentation for an undocumented constructor because dartdoc doesn't do so. I think it also makes sense to not do so just from the perspective that the documentation for the class isn't describing the constructor.

Just my thoughts on this, specifically:

> [...] the documentation for the class isn't describing the constructor.

If the class has no more than a single public constructor (usually happens to be unnamed, so that may be a point for this issue - we could, _maybe_ do a better job of handling named constructors too, see the image below), then the author probably doesn't care about documenting the constructor differently (I know I don't), because the constructor is not important at this point, only what the class represents. So I'd still be in favour of adding this exception to our tooling, and I'd even vote for Dartdoc to still maintain the current state. 

This would also be inconsistent with `public_member_api_docs` so I'd suggest if we do look into this, that we add this special-case for it too.

But yes, if we do have the docs for the constructor, then accessing the class documentation is hard, and we should definitely address that. FWIW, I like the idea of hovering `var`, `final` or `const`, but that would clash with https://github.com/dart-lang/sdk/issues/50085 if it is ever implemented, so we may need to think a bit more here.

<img width="600" height="306" alt="Image" src="https://github.com/user-attachments/assets/12979417-5061-4039-a834-4a96b23c055c" />

---

@DanTup, I was thinking of your example above:

<img width="600" height="306" alt="Image" src="https://user-images.githubusercontent.com/1078012/150164718-fde19eed-402b-430e-a0d5-787d76660989.png" />

See the part of the hover where we display the full constructor definition? Maybe we could add around it a folding section? So people could get to the docs easier? I'm not sure we can support that within VS Code, but considering all other MD on the hover, I suppose so. If that is done, we could possibly do the same to the constructor docs itself and place the class' docs after it as suggested above. If that is something you might like to address, maybe we can split my folding sections suggestion into a new issue (if there isn't one already).

**Comment by bwilkerson** [comment]
There is currently ongoing discussion about how to handle documentation comments for classes and constructors if we implement the  primary_constructors feature. I'd like to see how those discussions go so that we can make sure that any decision we make here is compatible with that feature.

I'm still not convinced that class documentation is helpful for explaining a constructor, but I can see your point about small classes being less likely to have a major distinction. If I think about a class like `Point` then maybe the class documentation _is_ sufficient. The question I haven't really convinced myself about is whether in cases like those it's better to display the class documentation, or to display nothing. If a user composes documentation for the class that isn't appropriate to a constructor, but didn't bother to document the constructor separately, do they really want the class documentation to show up when someone hovers over the constructor? If so, then why wouldn't DartDoc show the same docs?

> But yes, if we do have the docs for the constructor, then accessing the class documentation is hard, and we should definitely address that.

(I'm attempting here to try to get to the heart of the issue, not to minimize your opinion.) Why? For any other identifiers, hovering over them shows the documentation for the declaration being referenced. Why should constructors be any different? I understand that Dart has the unfortunate characteristic that the name of the so-called "unnamed" constructor is the same as the name of the class, but it's still defined to be the name of the constructor in an invocation of the constructor.

**Comment by FMorschel** [comment]
> If so, then why wouldn't DartDoc show the same docs?

I believe that, although the consistency is good, IMO, the purpose of DartDoc is to be accurate with the documentation provided. And I was considering that, the whole time I've used dartdoc, _if_ I ever went to see a constructor docs, I (1) found it through the class page, so I've just seen the class doc (2) did so to understand something about that particular constructor (and only had to do so when more than one was available).

On the other hand, the hover on the IDE (courtesy of DAS) is intended to help the user of the code, to easily find _relevant_ information about it, so it can be acted upon. If there is some information "missing", then the user has to try and navigate somewhere to find it, which indicates the tooling either can't or isn't doing something to help avoid that.

> > But yes, if we do have the docs for the constructor, then accessing the class documentation is hard, and we should definitely address that.
> 
> Why? For any other identifiers, hovering over them shows the documentation for the declaration being referenced. Why should constructors be any different? I understand that Dart has the unfortunate characteristic that the name of the so-called "unnamed" constructor is the same as the name of the class, but it's still defined to be the name of the constructor in an invocation of the constructor.

If I have a class (with multiple constructors, again, I see no sense in discussing this with a single constructor) and I find myself looking at an instance creation, I may want to either understand _what_ (class) is being created or _how_ (constructor) it is being created. For example, if I have a _controller_ of sorts, some of the time I may be interested to know _what it controls and how_ and not necessarily _how are these parameters being considered to create this controller_.

**Comment by benthillerkus** [comment]
For me the current situation is a big pain point when working with Flutter.

In Flutter, Widgets are typically const constructible data classes. If there aren't multiple named constructors, there's no value in documenting the constructor because it doesn't do anything interesting and has no meaning of its own.
However since Flutter `build` methods work by nesting Widget constructor calls without ever assigning them to a variable, you will never actually see class documentation without using "go-to-definition". You will only see the hover documentation of the constructor.

Just displaying the class documentation as a fallback if no documentation for the constructor is provided would be incredibly helpful.

A hack solution for the current situation could be to just document the constructors instead of the classes, but this would make the generated dart doc html worse to use and referencing other Widgets in the documentation becomes awkward since instead of mentioning `[MyCustomWidget]` in the markdown, I'd have to write `[MyCustomWidget.new]`.

> do they really want the class documentation to show up when someone hovers over the constructor?

I can't imagine having no documentation being preferable to having some?

> If so, then why wouldn't DartDoc show the same docs?

I'd echo what @FMorschel said: DartDoc is about accuracy and you access the documentation differently. In a generated DartDoc you'd typically find the class documentation before finding the constructor documentation.

**Comment by FMorschel** [comment]
Once we fix https://github.com/dart-lang/sdk/issues/62115, there will be a way to see the class docs on named constructors. 

But that would still be an issue on dot shorthand invocations, so we should consider that too when we come back to this (when it gets unblocked).
